Source/Scene/VoxelPrimitive.js

```diff
@@ -0,0 +1,2514 @@
+
+    shaderBuilder.addFunction(
+      sumAttributesFunctionId,
+      sumAttributesFunctionDeclaration,
+      ShaderDestination.FRAGMENT
+    );
+
+    shaderBuilder.addFunctionLines(sumAttributesFunctionId, [
+      `${attributeStructName} attributes;`,
+    ]);
+    for (let i = 0; i < attributeLength; i++) {
+      const name = names[i];
+      shaderBuilder.addFunctionLines(sumAttributesFunctionId, [
+        `${attributeFieldName}.${name} = attributesA.${name} + attributesB.${name};`,
+      ]);
+    }
+    shaderBuilder.addFunctionLines(sumAttributesFunctionId, [
+      `return attributes;`,
+    ]);
+  }
+
+  // mixAttributes
+  {
+    const mixAttributesFunctionId = "mixAttributes";
+    const mixAttributesFunctionName = "mixAttributes";
+    const mixAttributesFieldAttributesA = "attributesA";
+    const mixAttributesFieldAttributesB = "attributesB";
+    const mixAttributesFieldMixAmount = "mixFactor";
+    shaderBuilder.addFunction(
+      mixAttributesFunctionId,
+      `${attributeStructName} ${mixAttributesFunctionName}(${attributeStructName} ${mixAttributesFieldAttributesA}, ${attributeStructName} ${mixAttributesFieldAttributesB}, float ${mixAttributesFieldMixAmount})`,
+      ShaderDestination.FRAGMENT
+    );
+
+    shaderBuilder.addFunctionLines(mixAttributesFunctionId, [
+      `${attributeStructName} attributes;`,
+    ]);
+    for (let i = 0; i < attributeLength; i++) {
+      const name = names[i];
+      shaderBuilder.addFunctionLines(mixAttributesFunctionId, [
+        `attributes.${name} = mix(${mixAttributesFieldAttributesA}.${name}, ${mixAttributesFieldAttributesB}.${name}, ${mixAttributesFieldMixAmount});`,
+      ]);
+    }
+    shaderBuilder.addFunctionLines(mixAttributesFunctionId, [
+      `return attributes;`,
+    ]);
+  }
+
+  // setMinMaxAttributes function
+  if (defined(minimumValues) && defined(maximumValues)) {
+    shaderBuilder.addDefine(
+      "HAS_MIN_MAX",
+      undefined,
+      ShaderDestination.FRAGMENT
+    );
+    const minMaxAttributesFunctionId = "setMinMaxAttributes";
+    const minMaxAttributesFunctionName = "setMinMaxAttributes";
+    shaderBuilder.addFunction(
+      minMaxAttributesFunctionId,
+      `void ${minMaxAttributesFunctionName}(inout ${voxelStructName} ${voxelFieldName})`,
+      ShaderDestination.FRAGMENT
+    );
+    for (let i = 0; i < attributeLength; i++) {
+      const name = names[i];
+      const type = types[i];
+      const componentCount = MetadataType.getComponentCount(type);
+      for (let j = 0; j < componentCount; j++) {
+        const minimumValue = minimumValues[i][j];
+        const maximumValue = maximumValues[i][j];
+
+        // glsl needs to have `.0` at the end of whole numbers floats.
+        let minimumValueString = minimumValue.toString();
+        if (minimumValueString.indexOf(".") === -1) {
+          minimumValueString = `${minimumValue}.0`;
+        }
```

Comment:
Might be handy to have a helper function for this, I'd imagine that could get reused easily in other metadata code.